modules/core/src/main/java/org/apache/ignite/internal/processors/query/GridQueryProcessor.java

```diff
@@ -340,27 +348,121 @@ public void onCacheReconnect() throws IgniteCheckedException {
+                Map<UUID, Serializable> nodeSpecific = data.nodeSpecificData();
+
+                for (UUID nodeId : nodeSpecific.keySet()) {
+                    if (nodeSpecific.get(nodeId) instanceof Map)
+                        checkInlineSizes(indexesInlineSize, (Map<String, Integer>)nodeSpecific.get(nodeId), nodeId);
+                }
+            }
+        }
+    }
+
+    /**
+     *
+     * @return Information about secondary indexes inline size. Key is a full index name, value is a effective inline size.
+     * @see GridQueryIndexing#secondaryIndexesInlineSize()
+     */
+    public Map<String, Integer> secondaryIndexesInlineSize() {
+        return idx != null ? idx.secondaryIndexesInlineSize() : Collections.emptyMap();
+    }
+
+    /**
+     * Checks that server node with {@code nodeId} is supported {@link IgniteFeatures#CHECK_INDEX_INLINE_SIZES}.
+     *
+     * @param nodeId Node id.
+     * @return {@code true} if node with {@code nodeId} is server node and the node supports feature, {@code false}
+     * otherwise.
+     */
+    private boolean nodeSupportedCheckIndexInlineSize(UUID nodeId) {
+        IgnitePredicate<ClusterNode> nodeFilter = n -> !n.isClient() && !n.isDaemon() && nodeId.equals(n.id());
+
+        return IgniteFeatures.allNodesSupport(ctx, CHECK_INDEX_INLINE_SIZES, nodeFilter);
+    }
+
+    /**
+     * Compares indexes inline size on remote and local and fires warn message, if difference found.
+     *
+     * @param local Information about indexes inline size on local node.
+     * @param remote Information about indexes inline size on remote node.
+     * @param remoteNodeId Remote node id.
+     */
+    private void checkInlineSizes(Map<String, Integer> local, Map<String, Integer> remote, UUID remoteNodeId) {
+        SB sb = new SB();
+
+        if (log.isDebugEnabled())
+            log.debug("Check inline sizes on remote node with node id: " + remoteNodeId + ". Local: " + local + ", remote: " + remote);
+
+        for (String indexFullname : local.keySet()) {
+            if (remote.containsKey(indexFullname)) {
+                int localInlineSize = local.get(indexFullname);
+                int remoteInlineSize = remote.get(indexFullname);
+
+                if (localInlineSize != remoteInlineSize)
+                    sb.a(indexFullname).a("(").a(localInlineSize).a(",").a(remoteInlineSize).a(")").a(",");
+            }
+        }
+
+        if (sb.length() > 0) {
+            sb.setLength(sb.length() - 1);
+
+            log.warning(String.format(INLINE_SIZES_DIFFER_WARN_MSG_FORMAT, remoteNodeId, sb));
+        }
+    }
+
+    /**
+     * @return Serialiazible information about secondary indexes inline size.
+     * @see #secondaryIndexesInlineSize()
+     */
+    private Serializable collectSecondaryIndexesInlineSize() {
+        Map<String, Integer> map = secondaryIndexesInlineSize();
+
+        return map instanceof Serializable ? (Serializable)map : new HashMap<>(map);
```

Comment:
not serializable map ? plz clarify here ?

Reply:
Map interface doesn't extend java.io.Serializable interface. We could get here some custom no serializable implementation